File: GFG POTD/022-September-2025/017-decode-the-string.java
// https://www.geeksforgeeks.org/problems/decode-the-string2444/1/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(N)
    static String decodeString(String s) {
        Stack<Character> stk = new Stack<>();
        
        for (char ch : s.toCharArray()) {
            if (ch == ']') {
                // get the substring b/w [...]
                StringBuilder substr = new StringBuilder();
                while (!stk.isEmpty() && stk.peek() != '[') {
                    substr.append(stk.pop());
                }
                substr.reverse();
                
                if (!stk.isEmpty() && stk.peek() == '[') stk.pop();  // remove ']'
                
                int times = getTimes(stk);
                
                StringBuilder expanded = new StringBuilder();
                for (int i = 0; i < times; i ++) {
                    expanded.append(substr);
                }
                
                for (char c : expanded.toString().toCharArray()) {
                    stk.push(c);
                }
                
            } else {
                stk.push(ch);
            }
        }
        
        StringBuilder result = new StringBuilder();
        while (!stk.isEmpty()) {
            result.append(stk.pop());
        }
        result.reverse();
        return result.toString();
    }
    
    static private int getTimes(Stack<Character> stk) {
        int result = 0, base = 1;
        while (!stk.isEmpty() && isDigit(stk.peek())) {
            result = (stk.peek() - '0') * base + result;
            base  *= 10;
            stk.pop();
        }
        return result;
    }
    
    static private boolean isDigit(char ch) {
        return (ch >= '0' && ch <= '9');
    }
}

File: GFG POTD/022-September-2025/017-decode-the-string.cpp
// https://www.geeksforgeeks.org/problems/decode-the-string2444/1/




#include <bits/stdc++.h>
using namespace std;

class Solution {
  private:
    bool isDigit(char ch) {
        return (ch >= '0' && ch <= '9');
    }
    
    int getDigit(stack<char> &stk) {
        int result = 0, base = 1;
        while (!stk.empty() && isDigit(stk.top())) {
            result = (stk.top() - '0') * base + result;
            stk.pop();
            base *= 10;
        }
        return result;
    }
  public:
    // TC : O(N)
    // SC : O(N)
    string decodedString(string &s) {
        stack<char> stk;
        for (char &ch : s) {
            if (ch == ']') {
                // collect substring inside [...]
                string newStr;
                while (!stk.empty() && stk.top() != '[') {
                    newStr += stk.top();
                    stk.pop();
                }
                reverse(begin(newStr), end(newStr));
                
                // pop the '['
                if (!stk.empty() && stk.top() == '[') stk.pop();
                
                int times = getDigit(stk);
                
                string extended;
                for (int i = 0; i < times; i ++) {
                    extended += newStr;
                }

                for (char &c : extended) {
                    stk.push(c);
                }
            } else {
                stk.push(ch);
            }
        }
        
        string result;
        while (!stk.empty()) {
            result += stk.top();
            stk.pop();
        }
        reverse(begin(result), end(result));
        return result;
    }
};